ssl/ssl_lib.c

```diff
@@ -3823,6 +3823,169 @@ static int ssl_session_cmp(const SSL_SESSION *a, const SSL_SESSION *b)
     return memcmp(a->session_id, b->session_id, a->session_id_length);
 }
 
+SSL_CTX *SSL_CTX_dup(OSSL_LIB_CTX *libctx, SSL_CTX *source,
+                     const char *propq, const SSL_METHOD *meth)
+{
+    SSL_CTX *ret = NULL;
+    int ok = 0;
+
+    if (source == NULL)
+        goto exit;
+
+    ret = SSL_CTX_new_ex(libctx, propq, meth);
+    if (ret) {
+        if (meth != NULL)
+            ret->cnf->method = meth;
+        else
+            ret->cnf->method = source->cnf->method;
+        if (source->cnf->cipher_list) {
+            sk_SSL_CIPHER_free(ret->cnf->cipher_list);
+            ret->cnf->cipher_list = sk_SSL_CIPHER_dup(source->cnf->cipher_list);
+        }
+        if (source->cnf->cipher_list_by_id) {
+            sk_SSL_CIPHER_free(ret->cnf->cipher_list_by_id);
+            ret->cnf->cipher_list_by_id = sk_SSL_CIPHER_dup(source->cnf->cipher_list_by_id);
+        }
+        if (source->cnf->tls13_ciphersuites) {
+            sk_SSL_CIPHER_free(ret->cnf->tls13_ciphersuites);
+            ret->cnf->tls13_ciphersuites = sk_SSL_CIPHER_dup(source->cnf->tls13_ciphersuites);
+        }
+        SSL_CTX_set1_cert_store(ret, source->cnf->cert_store);
+        ret->cnf->session_cache_size = source->cnf->session_cache_size;
```

Comment:
Why not just do: `ret->cnf = source->cnf` and then deal with the _special_ fields?

The only down side would be splitting the frees from the duplications.

Reply:
I thought about this but then it is kind of weird if you have one SSL_CTX instance influences the other one when setting some configurations. 
Another issue that came to my mind was thread safety. That would mean for every setting API I need to lock.

What you think?

Reply:
Locking hardly seems relevant since you're not doing it currently....

Reply:
There are so many fields that it is probably better to do what pauli suggested... Note that this may still be prone to error when someone adds a new pointer field..
SO something like ret->cnf = source->cnf;
and then quite a lot of NULL sets ( In case we fail later)

ret->cnf->XXX = NULL;
ret->cnf->... = NULL;
ret->cnf->YYY = NULL;

ret->cnf->XXX = ....
if (ret->cnf->XXX == NULL) {
  Need to make sure all the XXX are cleaned up
  goto err;
}

Reply:
Will do, I just wait on @mattcaswell response if he finds this PR useful or not.

Reply:
I've created a new pull request which can be found at https://github.com/openssl/openssl/pull/23241.
The SSL_CTX config will be shared and access of the SSL_CTX config is locked.